Python/compile.c

```diff
@@ -1946,17 +1948,24 @@ compiler_call_exit_with_nones(struct compiler *c) {
 static int
 compiler_add_yield_from(struct compiler *c, int await)
 {
-    NEW_JUMP_TARGET_LABEL(c, start);
-    NEW_JUMP_TARGET_LABEL(c, resume);
+    NEW_JUMP_TARGET_LABEL(c, send);
+    NEW_JUMP_TARGET_LABEL(c, fail);
     NEW_JUMP_TARGET_LABEL(c, exit);
 
-    USE_LABEL(c, start);
+    USE_LABEL(c, send);
     ADDOP_JUMP(c, SEND, exit);
-
-    USE_LABEL(c, resume);
+    // Set up a virtual try/except to handle when StopIteration is raised during
+    // a close or throw call. The only way YIELD_VALUE raises if they do!
+    ADDOP_JUMP(c, SETUP_FINALLY, fail);
+    RETURN_IF_FALSE(compiler_push_fblock(c, TRY_EXCEPT, send, NO_LABEL, NULL));
```

Comment:
There can be no call to `compiler_unwind_fblock` before this block is popped, so we don't need to push a block here.